codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoderForBuffer.java

```diff
@@ -0,0 +1,662 @@
+ * </pre>
+ *
+ * <h3>Frame detection</h3>
+ * <p>
+ * Generally frame detection should be handled earlier in the pipeline by adding a
+ * {@link DelimiterBasedFrameDecoder}, {@link FixedLengthFrameDecoder}, {@link LengthFieldBasedFrameDecoder},
+ * or {@link LineBasedFrameDecoder}.
+ * <p>
+ * If a custom frame decoder is required, then one needs to be careful when implementing
+ * one with {@link ByteToMessageDecoderForBuffer}. Ensure there are enough bytes in the buffer for a
+ * complete frame by checking {@link Buffer#readableBytes()}. If there are not enough bytes
+ * for a complete frame, return without modifying the reader index to allow more bytes to arrive.
+ * <p>
+ * To check for complete frames without modifying the reader index, use methods like {@link Buffer#getInt(int)}.
+ * One <strong>MUST</strong> use the reader index when using methods like {@link Buffer#getInt(int)}.
+ * For example calling <tt>in.getInt(0)</tt> is assuming the frame starts at the beginning of the buffer, which
+ * is not always the case. Use <tt>in.getInt(in.readerIndex())</tt> instead.
+ * <h3>Pitfalls</h3>
+ * <p>
+ * Be aware that sub-classes of {@link ByteToMessageDecoderForBuffer} <strong>MUST NOT</strong>
+ * annotated with {@link @Sharable}.
+ */
+public abstract class ByteToMessageDecoderForBuffer extends ChannelHandlerAdapter {
+
+    /**
+     * Cumulate {@link Buffer}s by merge them into one {@link Buffer}'s, using memory copies.
+     */
+    public static final Cumulator MERGE_CUMULATOR = (alloc, cumulation, in) -> {
+        if (cumulation.readableBytes() == 0) {
+            // If cumulation is empty and input buffer is contiguous, use it directly
+            cumulation.close();
+            return in;
+        }
+        // We must close input Buffer in all cases as otherwise it may produce a leak if writeBytes(...) throw
+        // for whatever close (for example because of OutOfMemoryError)
+        try (in) {
+            final int required = in.readableBytes();
+            if (required > cumulation.writableBytes() || cumulation.readOnly()) {
+                return expandCumulation(alloc, cumulation, in);
+            }
+            cumulation.writeBytes(in);
+            return cumulation;
+        }
+    };
+
+    /**
+     * Cumulate {@link Buffer}s by add them to a {@link CompositeBuffer} and so do no memory copy whenever possible.
+     * Be aware that {@link CompositeBuffer} use a more complex indexing implementation so depending on your use-case
+     * and the decoder implementation this may be slower then just use the {@link #MERGE_CUMULATOR}.
+     */
+    public static final Cumulator COMPOSITE_CUMULATOR = (alloc, cumulation, in) -> {
+        if (cumulation.readableBytes() == 0) {
+            cumulation.close();
+            return in;
+        }
+        CompositeBuffer composite = null;
+        try {
+            if (CompositeBuffer.isComposite(cumulation)) {
+                composite = (CompositeBuffer) cumulation;
```

Comment:
If the write offset is not the same as the capacity, then we must use `split(write offset)` to cut off the unwritten end. Otherwise we'll run into trouble when we extend the composite buffer with new components. The composite buffer will be the result of the split, and the unwritten tail will be left in the cumulation, which we must then close.